cubi_tk/sodar/update_samplehseet.py

```diff
@@ -0,0 +1,628 @@
+        assay_cols = assay.columns.tolist()
+
+        isa_short_names = list(
+            map(lambda x: orig_col_name(isa_regex.sub(r"\2", x)), study_cols + assay_cols)
+        )
+        isa_long_names = list(map(orig_col_name, study_cols + assay_cols))
+        isa_names_unique = study_cols + assay_cols
+        isa_table = ["study"] * len(study_cols) + ["assay"] * len(assay_cols)
+
+        out = defaultdict(list)
+        for short, long, uniq, table in zip(
+            isa_short_names, isa_long_names, isa_names_unique, isa_table
+        ):
+            out[short].append((uniq, table))
+            out[long].append((uniq, table))
+
+        for col in ISA_NON_SETTABLE:
+            if col in out:
+                del out[col]
+
+        # do NOT retain defaultdict class
+        return dict(out)
+
+    def get_dynamic_columns(
+        self,
+        existing_names: Iterable[str],
+        isa_names: Iterable[str],
+    ) -> OrderedDict[str, str]:
+        dynamic_cols = OrderedDict()
+        re_format_names = re.compile(r"\{(.*?)}")
+        if self.args.dynamic_column:
+            for col, format_str in self.args.dynamic_column:
+                missing_deps = (
+                    set(re_format_names.findall(format_str))
+                    - set(existing_names)
+                    - set(dynamic_cols)
+                )
+                if missing_deps:
+                    raise ValueError(
+                        f"Dynamic column '{col}' depends on non-existing columns: {', '.join(missing_deps)}"
+                    )
+                if col not in isa_names:
+                    raise ValueError(f"Column '{col}' is not defined in the ISA samplesheet.")
+                dynamic_cols[col] = format_str
+        elif sheet_default_config[self.args.defaults]["dynamic_columns"]:
+            dynamic_cols = sheet_default_config[self.args.defaults]["dynamic_columns"]
+            # Hardcoded dep check for defaults: see if 'Library Name' is actually defined
+            if "Library Name" in dynamic_cols and "Library Name" not in isa_names:
+                logger.warning(
+                    'Skipping "Library Name" dynamic column, as it is not used in the ISA samplesheet.'
+                )
+                del dynamic_cols["Library Name"]
+
+        # #Sample Name needs to be set before other assay columns, so ensure it goes first
+        # if 'Sample Name' in dynamic_cols:
+        #     dynamic_cols.move_to_end('Sample Name', last=False)
+        return dynamic_cols
+
+    def collect_sample_data(
+        self,
+        isa_names: dict[str, list[tuple[str, str]]],
+        sample_field_mapping: dict[str, str],
+    ) -> pd.DataFrame:
+        ped_mapping = sheet_default_config[self.args.defaults]["ped_to_sampledata"]
+        if self.args.ped_field_mapping:
+            allowed_sample_col_values = [*sample_field_mapping] + [*isa_names]
+            for ped_col, sample_col in self.args.ped_field_mapping:
```

Comment:
Not sure about the types here, the value is a str in case of sample_field_mapping but a list[...] in case of isa_names

Reply:
This *should* generate (addable) lists of the dict keys, which are both str. Is there a better way to achieve this?

Reply:
For me, I think doing `allowed_sample_col_values = sample_field_mapping.keys() | isa_names.keys()` would suffice to be a little bit more explicit/descriptive.
(also just noticed that this file is called `update_samplehseet`, we should fix that)